[FEAT] Add support for Enterprise Cost Centers

vmvarela opened the pull request:
<!-- Please refer to our contributing docs for any questions on submitting a pull request -->
<!-- Issues are required for both bug fixes and features. -->
Resolves #2739 

----

### Before the change?
<!-- Please describe the current behavior that you are modifying. -->

* No support for enterprise cost centers resources or data

### After the change?
<!-- Please describe the behavior or changes that are being added by this PR. -->

* Two new resources and data sources (github_enterprise_cost_center and github_enterprise_cost_centers)

NOTE: This API (billing) has no support for APP or fine-grained tokens.

### Pull request checklist
- [x] Schema migrations have been created if needed ([example](https://github.com/integrations/terraform-provider-github/pull/2820/files))
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?
<!-- If this introduces a breaking change make sure to note it here any what the impact might be -->

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes
- [x] No

----

deiga commented:
We're trying to move towards using Context-aware provider functions, could you update all schemas to use those? So ReadContext etc

joraff commented:
I built and tested your branch @vmvarela and it works great! I was able to read, import, and create cost centers and resources using your examples and a few of my own. Thank you for the work.

deiga commented on github/data_source_github_enterprise_cost_center_test.go:
```diff
@@ -0,0 +1,73 @@
+		resource.TestCheckTypeSetElemAttr("data.github_enterprise_cost_center.test", "repositories.*", testEnterpriseCostCenterRepository),
+		resource.TestCheckResourceAttr("data.github_enterprise_cost_center.test", "users.#", "1"),
+		resource.TestCheckTypeSetElemAttr("data.github_enterprise_cost_center.test", "users.*", users[0]),
+	)
```
Please inline this variable

deiga commented on github/data_source_github_enterprise_cost_center_test.go:
```diff
@@ -0,0 +1,73 @@
+		resource.TestCheckTypeSetElemAttr("data.github_enterprise_cost_center.test", "organizations.*", testEnterpriseCostCenterOrganization),
+		resource.TestCheckResourceAttr("data.github_enterprise_cost_center.test", "repositories.#", "1"),
+		resource.TestCheckTypeSetElemAttr("data.github_enterprise_cost_center.test", "repositories.*", testEnterpriseCostCenterRepository),
+		resource.TestCheckResourceAttr("data.github_enterprise_cost_center.test", "users.#", "1"),
```
issue: Since the count of users depends on the input list length, this check should be using the len of the usersList

deiga commented on github/data_source_github_enterprise_cost_center_test.go:
```diff
@@ -0,0 +1,73 @@
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { skipUnlessMode(t, enterprise) },
+		Providers: testAccProviders,
```
```suggestion
		ProviderFactories: providerFactories,
```

deiga commented on github/data_source_github_enterprise_cost_centers.go:
```diff
@@ -0,0 +1,110 @@
+		state = &s
+	}
+
+	ctx = context.WithValue(ctx, ctxId, fmt.Sprintf("%s/cost-centers", enterpriseSlug))
```
```suggestion
```

deiga commented on github/data_source_github_enterprise_cost_centers.go:
```diff
@@ -0,0 +1,110 @@
+	if state != nil {
+		stateStr = *state
+	}
+	d.SetId(fmt.Sprintf("%s/%s", enterpriseSlug, stateStr))
```
```suggestion
	d.SetId(buildTwoPartID(enterpriseSlug, stateStr))
```

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceGithubEnterpriseCostCenterImport,
+		},
+
```
Please add a top-level `Description` attribute

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+	sort.Strings(toAdd)
+	sort.Strings(toRemove)
+	return toAdd, toRemove
+}
```
question: This looks quite similar to `func setChanges` which exists, could that be re-used?

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+
+	if hasCostCenterAssignmentsConfigured(d) {
+		// Ensure we operate on fresh API state before mutations.
+		current, _, err := client.Enterprise.GetCostCenter(ctx, enterpriseSlug, cc.ID)
```
question: Why is this necessary in `Create`?

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+	cc, _, err := client.Enterprise.GetCostCenter(ctx, enterpriseSlug, costCenterID)
+	if err != nil {
+		if is404(err) {
+			// If the API starts returning 404 for archived cost centers, we remove it from state.
```
Please add `tflog.Warn` with mention of the removal

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+	enterpriseSlug := d.Get("enterprise_slug").(string)
+	costCenterID := d.Id()
+
+	ctx = context.WithValue(ctx, ctxId, fmt.Sprintf("%s/%s", enterpriseSlug, costCenterID))
```
```suggestion
```

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+		return diag.FromErr(err)
+	}
+	if strings.EqualFold(cc.GetState(), "deleted") {
+		return diag.FromErr(fmt.Errorf("cannot update cost center %q because it is archived", costCenterID))
```
```suggestion
		return diag.Errorf("cannot update cost center %q because it is archived", costCenterID)
```

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+	enterpriseSlug := d.Get("enterprise_slug").(string)
+	costCenterID := d.Id()
+
+	ctx = context.WithValue(ctx, ctxId, fmt.Sprintf("%s/%s", enterpriseSlug, costCenterID))
```
```suggestion
```

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+		return nil, fmt.Errorf("invalid import specified: supplied import must be written as <enterprise_slug>/<cost_center_id>")
+	}
+
+	enterpriseSlug, costCenterID := parts[0], parts[1]
```
Let's use a `:` as the import ID separator, then you can use `parseTwoPartID`

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,452 @@
+		out = append(out, s)
+	}
+	return out
+}
```
issue: This is a duplicate of `flattenStringList`, please use that instead
```suggestion
```

deiga reviewed (changes requested):
Half-way through. Please consider if my comments could apply to other places in your changes as well, I wasn't able to comment on every duplicate thing

vmvarela reviewed (commented):
Thank you for the thorough review @deiga! I've addressed all the feedback. Here's a summary of the changes:

**Implemented:**
- ✅ Added top-level `Description` attribute to all resources and data sources
- ✅ Removed `ctx = context.WithValue(ctx, ctxId, ...)` pattern from all files
- ✅ Replaced `log.Printf` with `tflog.Info` and added `tflog.Warn` for 404 removals
- ✅ Used `diag.Errorf` instead of `diag.FromErr(fmt.Errorf(...))`
- ✅ Create and Update no longer call Read - computed fields are set directly
- ✅ Extracted anonymous retry functions to named functions (`retryCostCenterAddResources`, `retryCostCenterRemoveResources`)
- ✅ Changed import ID separator from `/` to `:` and using `parseTwoPartID`
- ✅ Replaced `stringSliceToAnySlice` with existing `flattenStringList`
- ✅ Used `http.StatusNotFound` and other constants instead of magic numbers
- ✅ Moved `is404` to `util.go`
- ✅ Updated tests to use `providerFactories` and `skipUnlessEnterprise(t)`
- ✅ Added `owner = var.enterprise_slug` to example provider config
- ✅ Used `buildTwoPartID` in data source
- ✅ Inlined check variables in tests

**Regarding questions:**
- The `resources` attribute provides the raw API view while `users`, `organizations`, `repositories` are filtered convenience attributes. I kept both for flexibility and debugging purposes.
- `diffStringSlices` differs from `setChanges` - it compares string slices directly while `setChanges` works with Terraform state changes. They serve different purposes.
- In Create, we only fetch from API after assignments are configured to populate computed fields, since we're not calling Read anymore.

All tests pass and the code compiles without errors.

deiga commented on github/data_source_github_enterprise_cost_centers_test.go:
```diff
@@ -0,0 +1,72 @@
+			Config: config,
+			Check: resource.ComposeTestCheckFunc(
+				resource.TestCheckResourceAttr("data.github_enterprise_cost_centers.test", "state", "active"),
+				testAccCheckEnterpriseCostCentersListContains("github_enterprise_cost_center.test", "data.github_enterprise_cost_centers.test"),
```
suggestion: You could replace this with 
```suggestion
				resource.TestCheckTypeSetElemAttr("data.github_enterprise_cost_centers.test", "cost_centers.*.*", github_enterprise_cost_center.test.id),
```

deiga commented on github/resource_github_enterprise_cost_center_test.go:
```diff
@@ -0,0 +1,110 @@
+		resource.TestCheckResourceAttr("github_enterprise_cost_center.test", "state", "active"),
+		resource.TestCheckResourceAttr("github_enterprise_cost_center.test", "users.#", "1"),
+		resource.TestCheckTypeSetElemAttr("github_enterprise_cost_center.test", "users.*", user),
+	)
```
Please inline these

deiga commented on github/resource_github_enterprise_cost_center_test.go:
```diff
@@ -0,0 +1,110 @@
+					return fmt.Sprintf("%s:%s", testAccConf.enterpriseSlug, rs.Primary.ID), nil
+				},
+			},
+		},
```
Please split this into multiple tests instead of doing all operations in one test

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,451 @@
+			if err != nil {
+				return diag.FromErr(err)
+			}
+			currentResources = cc.Resources
```
question: Why would this be necessary? Doesn't `terraform plan` already call `Read` and `d.GetChange` would already get you old and new values

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,451 @@
+		state = "active"
+	}
+	_ = d.Set("state", state)
+	_ = d.Set("azure_subscription", cc.GetAzureSubscription())
```
Please return any errors from setting the attributes

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,451 @@
+func syncEnterpriseCostCenterAssignments(ctx context.Context, d *schema.ResourceData, client *github.Client, enterpriseSlug, costCenterID string, currentResources []*github.CostCenterResource) diag.Diagnostics {
+	desiredUsers := expandStringSet(getStringSetOrEmpty(d, "users"))
+	desiredOrgs := expandStringSet(getStringSetOrEmpty(d, "organizations"))
+	desiredRepos := expandStringSet(getStringSetOrEmpty(d, "repositories"))
```
question: I'm wondering if `currentResources` is ever actually useful.
I'm thinking that `d.hasChange` and `d.getChange` should always already produce the required data, right?
```suggestion
	var desiredRepos, currentRepos *schema.Set
	if d.hasChange("repositories") {
		currentRepos, desiredRepos := setChanges(data.GetChange("repositories"))
	} else {
		currentRepos := schema.NewSet(schema.HashString, []any{})
		desiredRepos := schema.NewSet(schema.HashString, []any{})
	}
	toRemoveRepos := currentRepos.Difference(desiredRepos).List()
	toAddRepos := desiredRepos.Difference(currentRepos).List()
```

deiga reviewed (changes requested):
Please pay attention to what I'm asking form you and make sure to make similar changes in other places where they might apply.

Consider these also for your other PRs

vmvarela commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,451 @@
+
+	d.SetId(cc.ID)
+
+	if hasCostCenterAssignmentsConfigured(d) {
```
Thanks for the suggestion! I considered using MinItems: 1, but I think the current approach is more flexible for users.

With MinItems: 1, users wouldn't be able to explicitly set users = [] to clear all assignments - they'd get a validation error. The current design allows three states:

- Attribute not set (nil) → don't manage assignments
- Empty set [] → remove all assignments
- Non-empty set → manage specific assignments

The hasCostCenterAssignmentsConfigured() check is only used in Create to determine if we need to sync assignments after creating the cost center. It distinguishes between "user didn't configure any assignments" vs "user explicitly wants an empty set".

If you prefer the MinItems: 1 approach and accept that users must omit the attribute entirely (rather than set it empty) to have no assignments, I can make that change. Let me know your preference!

deiga commented on github/data_source_github_enterprise_cost_centers.go:
```diff
@@ -0,0 +1,98 @@
+	var state *string
+	if v, ok := d.GetOk("state"); ok {
+		s := v.(string)
+		state = &s
```
```suggestion
		state = github.Ptr(v.(string))
```

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,515 @@
+
+	if cc == nil || cc.ID == "" {
+		return diag.Errorf("failed to create cost center: missing id in response")
+	}
```
question: What are the situations where this condition would be true? If this is actually likely to happen, can we improve error message? A user who gets this wouldn't know what to do with it

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,515 @@
+	desiredOrgs := expandStringSet(getStringSetOrEmpty(d, "organizations"))
+	desiredRepos := expandStringSet(getStringSetOrEmpty(d, "repositories"))
+
+	currentUsers, currentOrgs, currentRepos := costCenterSplitResources(currentResources)
```
issue: `currentResources` will be always nil here and probably panic. Have you made sure to run the tests?

deiga commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,515 @@
+	users, organizations, repositories := costCenterSplitResources(cc.Resources)
+	sort.Strings(users)
+	sort.Strings(organizations)
+	sort.Strings(repositories)
```
since you're using `TypeSet` I'm not sure if sorting has any impact

deiga reviewed (approved):
LGTM, still need review from other maintainers

stevehipwell reviewed (changes requested):
Thanks for the PR @vmvarela, I've added some general comments.

But my main review comment is that the pattern you're attempting to use for `github_enterprise_cost_center` is mixing concerns. Could you please separate out the assignments from the `github_enterprise_cost_center` resource. You could create a `github_enterprise_cost_center_assignment` resource to keep the authoritative pattern if it needs to span all types; but `github_enterprise_cost_center_users`, `github_enterprise_cost_center_organizations` & `github_enterprise_cost_center_repositories` would be the preferred pattern. This way the provider logic is more consistent and we could support non-authoritative patterns in the future.

github-actions[bot] committed:
feat(utils): add shared helper functions for cost centers

Add utility functions needed for enterprise cost center resources:
- errIs404(): check if error is a GitHub 404 Not Found response
- errIsRetryable(): check if error is retryable (409, 5xx)
- expandStringSet(): convert schema.Set to []string
- chunkStringSlice(): split slice into chunks for batching

github-actions[bot] committed:
feat(cost-centers): add retry logic utilities for cost center operations

Add util_enterprise_cost_center.go with helper functions for managing
cost center assignments with proper retry logic and batching support.

github-actions[bot] committed:
feat(cost-centers): add github_enterprise_cost_center resource

Manages GitHub Enterprise cost center entities (create, read, update, archive).

Includes:
- Resource implementation with CRUD operations
- Acceptance tests
- Documentation

github-actions[bot] committed:
feat(cost-centers): add github_enterprise_cost_center_users resource

Authoritative management of user assignments to cost centers.

Includes:
- Resource implementation with batched add/remove operations
- Acceptance tests
- Documentation

github-actions[bot] committed:
feat(cost-centers): add github_enterprise_cost_center_organizations resource

Authoritative management of organization assignments to cost centers.

Includes:
- Resource implementation with batched add/remove operations
- Acceptance tests
- Documentation

github-actions[bot] committed:
feat(cost-centers): add github_enterprise_cost_center_repositories resource

Authoritative management of repository assignments to cost centers.

Includes:
- Resource implementation with batched add/remove operations
- Acceptance tests
- Documentation

github-actions[bot] committed:
feat(cost-centers): add data sources for enterprise cost centers

Add two data sources:
- github_enterprise_cost_center: retrieve a cost center by ID
- github_enterprise_cost_centers: list cost centers with optional state filter

Includes:
- Data source implementations
- Acceptance tests
- Documentation

github-actions[bot] committed:
feat(cost-centers): register cost center resources and data sources

- Register 4 resources in provider.go:
  - github_enterprise_cost_center
  - github_enterprise_cost_center_users
  - github_enterprise_cost_center_organizations
  - github_enterprise_cost_center_repositories
- Register 2 data sources in provider.go:
  - github_enterprise_cost_center
  - github_enterprise_cost_centers
- Add navigation links in website/github.erb

T_EDITOR=true git rebase --continue
t status

github-actions[bot] committed:
docs(cost-centers): add usage example

Add example Terraform configuration demonstrating how to:
- Create a cost center
- Assign users, organizations, and repositories
- Use data sources to query cost centers

github-actions[bot] committed:
fix(cost-centers): update go-github import to v82

Update import paths from go-github/v81 to go-github/v82 to match
the current version in upstream/main.

deiga commented on github/util.go:
```diff
@@ -302,6 +302,53 @@ func getTeamSlugContext(ctx context.Context, teamIDString string, meta any) (str
+
+// chunkStringSlice splits a slice into chunks of the specified max size.
+//
+//nolint:unparam // maxSize is parameterized for reusability across different contexts
```
Please remove this, it shouldn't be needed
```suggestion
```

deiga commented on github/util.go:
```diff
@@ -302,6 +302,53 @@ func getTeamSlugContext(ctx context.Context, teamIDString string, meta any) (str
+	}
+	list := set.List()
+	return expandStringList(list)
+}
```
issue: This function seems unnecessary, since in all 3 cases it's being used the `set` can't be `nil`. So, it seems to just be abstracting away the call `expandStringList(set.List())`

deiga commented on github/resource_github_enterprise_cost_center_users.go:
```diff
@@ -0,0 +1,201 @@
+		}
+	}
+
+	return resourceGithubEnterpriseCostCenterUsersRead(ctx, d, meta)
```
Please set any computed fields directly, instead of calling `Read`
```suggestion
	return nil
```

deiga commented on github/resource_github_enterprise_cost_center_organizations.go:
```diff
@@ -0,0 +1,193 @@
+			return diag.FromErr(err)
+		}
+		d.SetId(id)
+	}
```
issue: instead of this condition, please separate `Create` and `Update`

deiga commented on github/resource_github_enterprise_cost_center_repositories.go:
```diff
@@ -0,0 +1,193 @@
+	}
+}
+
+func resourceGithubEnterpriseCostCenterRepositoriesCreateOrUpdate(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
```
issue: Please split this into `Create` and `Update`

vmvarela replied:
Done

vmvarela committed:
fix: removed linter config, it shouldn't be needed

Co-authored-by: Timo Sand <timo.sand@iki.fi>

vmvarela committed:
fix(cost-centers): remove unnecessary expandStringSet function

Remove expandStringSet from util.go and replace usages with direct
expandStringList(set.List()) calls. The function was unnecessary since
schema.Set from d.Get() is never nil.

Resolves PR comments #1-2.

vmvarela committed:
fix(cost-centers): check deleted state in Read instead of Update

Move the archived/deleted state check from Update to Read function.
If the cost center is archived (deleted), it will be removed from
Terraform state during Read rather than blocking updates.

Resolves PR comment #3.

vmvarela committed:
fix(cost-centers): separate Create and Update for users resource

Split resourceGithubEnterpriseCostCenterUsersCreateOrUpdate into
separate Create and Update functions. Create only adds users,
Update handles the full diff. Both return nil instead of calling Read.

Resolves PR comments #4-5.

vmvarela committed:
fix(cost-centers): separate Create and Update for organizations resource

Split resourceGithubEnterpriseCostCenterOrganizationsCreateOrUpdate into
separate Create and Update functions. Create only adds organizations,
Update handles the full diff. Both return nil instead of calling Read.

Resolves PR comments #6-7.

vmvarela committed:
fix(cost-centers): separate Create and Update for repositories resource

Split resourceGithubEnterpriseCostCenterRepositoriesCreateOrUpdate into
separate Create and Update functions. Create only adds repositories,
Update handles the full diff. Both return nil instead of calling Read.

Resolves PR comments #8-9.

vmvarela committed:
fix(cost-centers): correct resource type strings in CheckDestroy functions

The API returns type strings as 'User', 'Org', and 'Repo' but the tests
were checking for lowercase 'user', 'organization', and 'repository'.
This fix ensures CheckDestroy properly detects remaining assignments.

vmvarela committed:
refactor(cost-centers): use type constants instead of magic strings

Add CostCenterResourceType constants (User, Org, Repo) to avoid
magic strings throughout the codebase. This prevents typos and
makes the code more maintainable.

Addresses review feedback from @deiga.

deiga reviewed (changes requested):
You didn't run tests or linter after rebasing. You need to adapt to the new changes

vmvarela committed:
fix(cost-centers): use terraform-plugin-testing imports in test files

Replace terraform-plugin-sdk/v2 test imports with terraform-plugin-testing
to fix flag redefinition conflict ('sweep' flag registered twice).
This aligns cost center tests with the rest of the codebase.

vmvarela committed:
fix(cost-centers): document generic chunk helper lint exception

Keep chunkStringSlice(items, maxSize) generic in util.go to avoid coupling with cost-center-specific constants.

Add nolint:unparam with explicit rationale because current call sites pass the same value, while preserving future reuse for other resources.

deiga commented on github/resource_github_enterprise_cost_center_test.go:
```diff
@@ -0,0 +1,120 @@
+							return "", err
+						}
+						return id, nil
+					},
```
Since we have pre-determined values in the Import ID we can directly use those
```suggestion
					ImportStateIdPrefix: testAccConf.enterpriseSlug + ":",
```

deiga commented on github/resource_github_enterprise_cost_center_users.go:
```diff
@@ -0,0 +1,225 @@
+	desiredUsersSet := d.Get("usernames").(*schema.Set)
+	toAdd := expandStringList(desiredUsersSet.List())
+
+	// Add users
```
Remove unnecessary comment
```suggestion
```

deiga commented on github/resource_github_enterprise_cost_center_users.go:
```diff
@@ -0,0 +1,225 @@
+	enterpriseSlug := d.Get("enterprise_slug").(string)
+	costCenterID := d.Get("cost_center_id").(string)
+
+	// Get current assignments from API
```
Remove unnecessary comment
```suggestion
```

deiga commented on github/resource_github_enterprise_cost_center_users.go:
```diff
@@ -0,0 +1,225 @@
+		}
+	}
+
+	// Add new users
```
Remove unnecessary comment
```suggestion
```

deiga reviewed (changes requested):
Please replace usage of `Check:` in `TestStep` by `ConfigStateCheck:` and use the new checks: https://developer.hashicorp.com/terraform/plugin/testing/acceptance-tests/state-checks/resource

vmvarela committed:
fix: address review - rename single-char variables and remove unnecessary comments

vmvarela committed:
fix: address review - rename maxResourcesPerRequest to maxCostCenterResourcesPerRequest

vmvarela commented on github/util_enterprise_cost_center.go:
```diff
@@ -0,0 +1,59 @@
+
+// Cost center resource management constants and retry functions.
+const (
+	maxResourcesPerRequest          = 50
```
Done

vmvarela commented on github/resource_github_enterprise_cost_center_users.go:
```diff
@@ -0,0 +1,225 @@
+		}
+	}
+
+	// Remove users no longer desired
```
Done

vmvarela commented on github/resource_github_enterprise_cost_center_users.go:
```diff
@@ -0,0 +1,225 @@
+	}
+	d.SetId(id)
+
+	// Get desired users from config
```
Done

vmvarela committed:
fix: address review - simplify import test with ImportStateIdPrefix

vmvarela committed:
fix: address review - add unit tests for errIs404, errIsRetryable, chunkStringSlice

vmvarela commented on github/util.go:
Done. Added unit tests for `errIs404`, `errIsRetryable`, and `chunkStringSlice` in `util_test.go`, covering all code paths (nil errors, plain errors, each HTTP status code, empty/exact/remainder/edge cases for chunking). The `nolint:unparam` on `chunkStringSlice` has been removed since the tests now call it with varying `maxSize` values.

vmvarela committed:
fix: address review - migrate cost center tests to ConfigStateChecks

deiga reviewed (approved):
Seems good. @stevehipwell go ahead :)

stevehipwell reviewed (changes requested):
@vmvarela I've just reviewed the data sources. I'll re-review the whole PR once you've rebased and are uisng the correct SDK version.

vmvarela committed:
fix(cost-centers): update go-github import from v82 to v83

vmvarela committed:
fix(cost-centers): add default 'all' value to state field in cost centers data source

stevehipwell reviewed (changes requested):
I'm assuming that you've used a consistent pattern so I've reviewed up to the first assignment resource.

vmvarela committed:
fix(cost-center): populate name on import via API lookup

The import function now calls GetCostCenter to populate the 'name'
field from the API response. Without this, the Required 'name' field
would be empty after import, causing recreate churn on the next plan.

Addresses review comment from @stevehipwell.

vmvarela committed:
fix(cost-center): align sub-resource IDs with main resource

Sub-resources (_users, _organizations, _repositories) now use a simple
cost_center_id as their Terraform ID, consistent with the main
github_enterprise_cost_center resource.

Also moves d.SetId() to after the API call succeeds, preventing
corrupt state if the add-resources call fails.

Tests updated to use ImportStateIdPrefix and read attributes from
state instead of parsing the now-simple ID.

Addresses review comments from @stevehipwell.

vmvarela committed:
fix(cost-center): validate no existing assignments on Create

Before adding resources to a cost center, check via API whether
the cost center already has resources of the managed type assigned.
If so, return an error asking the user to import or remove them
manually. This prevents silently clobbering pre-existing assignments.

vmvarela committed:
refactor(cost-center): simplify Update diff with map pattern

Replace the two-map diff (currentX + desiredX) with a single map
where false=remove and true=keep. Desired items not in the map are
added. This is shorter and avoids constructing a second map.

vmvarela committed:
fix(cost-center): Delete removes all linked resources from API

Instead of reading resource names from Terraform state (which may be
stale or incomplete), Delete now calls GetCostCenter to fetch the
current list of resources of the managed type from the API and removes
them all. Also handles 404 gracefully (cost center already gone).

vmvarela committed:
fix(cost-center): add missing sub-resources to website sidebar

The cost center sub-resources (organizations, repositories, users)
were registered in provider.go and had docs, but were accidentally
omitted from the website sidebar navigation.

vmvarela commented on github/resource_github_enterprise_cost_center_organizations.go:
```diff
@@ -0,0 +1,216 @@
+	costCenterID := d.Get("cost_center_id").(string)
+
+	organizationsSet := d.Get("organization_logins").(*schema.Set)
+	organizations := expandStringList(organizationsSet.List())
```
Addressed in 34031c6e — Delete now fetches all linked organizations from the API (not from state) before removing them. Same pattern applied to users and repositories sub-resources.

vmvarela commented:
Hey @deiga @stevehipwell — quick naming question before we finalize.

Would you think it's appropriate to add `billing` to the cost center resource names? This would better reflect their placement under the GitHub Billing API and help avoid ambiguity as the enterprise namespace grows.

The rename would be:

**Resources:**
- `github_enterprise_cost_center` → `github_enterprise_billing_cost_center`
- `github_enterprise_cost_center_users` → `github_enterprise_billing_cost_center_users`
- `github_enterprise_cost_center_organizations` → `github_enterprise_billing_cost_center_organizations`
- `github_enterprise_cost_center_repositories` → `github_enterprise_billing_cost_center_repositories`

**Data sources:**
- `github_enterprise_cost_center` → `github_enterprise_billing_cost_center`
- `github_enterprise_cost_centers` → `github_enterprise_billing_cost_centers`

The change is straightforward — just a rename across files, provider registration, and docs. Happy to keep the current names if you prefer; just wanted to get your take before tagging a release.

vmvarela committed:
fix: update go-github import to v84 and remove duplicate team helpers after rebase

vmvarela committed:
fix(cost-centers): use third-person verb in data source descriptions

Closes #51

deiga reviewed (changes requested):
We're close! :)

vmvarela committed:
chore(lint): exclude modernize newexpr rule for Ptr calls

The modernize/newexpr rule incorrectly suggests replacing github.Ptr(x)
with new(x), but new() takes a type not an expression — the auto-fix
generates invalid Go code. Exclude this specific rule pattern.

vmvarela committed:
fix(cost-centers): remove defensive nil check and use deleteResourceOn404AndSwallow304OtherwiseReturnError helper

vmvarela commented on github/resource_github_enterprise_cost_center.go:
```diff
@@ -0,0 +1,186 @@
+			})
+			d.SetId("")
+			return nil
+		}
```
Done

vmvarela commented on github/resource_github_enterprise_cost_center_organizations.go:
```diff
@@ -0,0 +1,231 @@
+			})
+			d.SetId("")
+			return nil
+		}
```
Done

vmvarela committed:
fix: undo .golangci.yml changes

vmvarela committed:
fix: after rebase

vmvarela commented:
Sorry @deiga I do not understand this change to milestone 6.13.

deiga commented:
@vmvarela We have reduced capacity on the maintainers team and want to release 6.12.0 with all the already merged changes. That's why anything we can't get merged this week has been moved to the next release

deiga commented:
Hey @vmvarela,

Would you have capacity to re-open this? We'd love to incorporate your changes :)
